accepted/0000-sbom-command.md

```diff
@@ -0,0 +1,461 @@
+
+
+## Motivation
+
+Finding and remediating vulnerabilities in open source projects is a critical component of securing the software supply chain. However, this requires that enterprises understand what OSS components are used across their infrastructure and applications. When new vulnerabilities are discovered, they need a complete inventory of their software assets in order to properly assess their exposure. Knowing about the critical [Log4j](https://www.cisa.gov/news-events/cybersecurity-advisories/aa21-356a#main) vulnerability doesn’t do you any good unless you can also pinpoint where in your organization you’re running the vulnerable code.
+
+SBOMs help to solve this problem by providing a standardized way to document the components that comprise a software application. A proper SBOM should tell you exactly which packages you have deployed and which versions of those packages you are using.
+
+Beyond the security benefit there may be a regulatory requirement to provide SBOMs in some sectors. In response to some recent, high-visibility attacks, The White House has issued an [Executive Order](https://www.whitehouse.gov/briefing-room/presidential-actions/2021/05/12/executive-order-on-improving-the-nations-cybersecurity/) which specifically includes directives which would make SBOMs a requirement for any vendor selling to the federal government.
+
+Adding SBOM generation to the tool which many developers are already using to manage their project dependencies eliminates any friction which may come from having to adopt/learn a separate tool. 
+
+
+## Detailed Explanation
+
+A new `sbom` command will be added to the npm CLI which will generate an SBOM for the current project. The SBOM will use the current project as the root and enumerate all of its dependencies (similar to the way `npm-ls` does) in one of the two supported SBOM formats. See the 
+[Example SBOMs](#example-sboms) section for sample CycloneDX and SPDX SBOM documents.
+
+Supported command options: 
+
+`--sbom-format` - SBOM format to use for output. Valid values are “spdx” or “cyclonedx”. In the future, the set of valid values can be expanded to select differents versions of the supported output formats (e.g. "cyclonedx1.4" vs "cyclonedx1.5")
+
+`--omit` - Dependency types to omit from generated SBOM. Valid values are “dev”, “optional”, and “peer” (can be set multiple times). By default, all development, optional, and peer dependencies will be included in the generated SBOM unless explicitly excluded.
+
+`--package-lock-only` - Constructs the SBOM based on the tree described by the _package-lock.json_, rather than the contents of _node_modules_. Defaults to _false_. If the _node_modules_ folder is not present, this flag will be required in order to generate an SBOM.
+ 
```

Comment:
I know the CycloneDX module supports this option, but given this is about integration with NPM itself, I can't help but feel like this "shortcut" should not be supported. IMO, if it's a feature native to the package manager, it should have one, and only one, "correct" option.

The only reason for supporting the `--package-lock-only` flag (in my mind), is that you want to generate a BOM in workflows where you can't or don't want to build the project. Say, a "generate BOMs for all projects in my org, without developers having to change their pipelines" kind of thing. That is a different lifecycle stage (pre-build or earlier) than analyzing the packages you have installed, which will be post-build.

If this option *must* be supported, then there should be ways for consumers to differentiate between how the BOM was generated. CycloneDX v1.5 comes with support for [lifecycle phases](https://cyclonedx.org/docs/1.5/json/#metadata_lifecycles_items_oneOf_i0_phase), which could be of use. There are also ways to communicate *how* components where identified via [evidence](https://cyclonedx.org/docs/1.5/json/#components_items_evidence_identity), which even includes [confidence](https://cyclonedx.org/docs/1.5/json/#components_items_evidence_identity_confidence).

For example, if components were identified solely via `package-lock.json`, the confidence should be lower than if they were identified on disk, from `node_modules`. The evidence should then also define the [method](https://cyclonedx.org/docs/1.5/json/#components_items_evidence_identity_methods_items_technique) as `manifest-analysis`.

Reply:
I personally don't feel strongly one way or the other about this particular option -- and can't really say whether the I-need-to-generated-an-SBOM-without-also-building-my-project use case is even a thing. I added it mainly because other SBOM tools support it (and it's pretty easy to implement within the npm CLI).

If the consensus is that this is not useful, I'm happy to remove it (and maybe reconsider in the future if real use cases arise). Explicitly identifying package-lock-only SBOMs also seems reasonable.

Reply:
of course it's a thing; any task should be able to be run in isolation. Including this option makes perfect sense.

Reply:
I would suggest adding support for CycloneDX lifecycle phases and including evidence and confidence as @nscuro suggested earlier. With an `--package-lock-only` option, lifecycles and evidence become much more important.

Reply:
For additional information about lifecycles and evidence, refer to:

- https://cyclonedx.org/guides/sbom/lifecycle_phases/
- https://cyclonedx.org/guides/sbom/evidence/

This all ties into SBOM quality which OWASP CycloneDX defines here: 
- https://cyclonedx.org/guides/sbom/bom/#sbom-quality